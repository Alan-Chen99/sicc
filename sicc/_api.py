from __future__ import annotations

import abc
import functools
from contextlib import AbstractContextManager
from dataclasses import dataclass
from typing import Any
from typing import Callable
from typing import Final
from typing import Protocol
from typing import TypeVar
from typing import Unpack
from typing import overload

from . import _functions as _f
from ._core import InstrBase
from ._core import InstrTypedWithArgs_api
from ._core import Label
from ._core import MVar
from ._core import Scope
from ._core import Value
from ._core import Var
from ._core import VarT
from ._core import can_cast_implicit_many
from ._core import get_type
from ._diagnostic import DebugInfo
from ._diagnostic import add_debug_info
from ._diagnostic import describe_fn
from ._diagnostic import must_use
from ._diagnostic import register_exclusion
from ._diagnostic import track_caller
from ._instructions import AddF
from ._instructions import AddI
from ._instructions import BlackBox
from ._instructions import Jump
from ._instructions import PredLE
from ._instructions import PredLT
from ._instructions import UnreachableChecked
from ._tracing import _CUR_SCOPE
from ._tracing import RawSubr
from ._tracing import ensure_label
from ._tracing import label
from ._tracing import mk_internal_label
from ._tracing import mk_mvar
from ._tracing import trace_if
from ._tracing import trace_to_raw_subr
from ._tree_utils import pytree
from ._utils import Cell
from ._utils import cast_unchecked
from ._utils import empty
from ._utils import empty_t
from ._utils import isinst
from ._utils import late_fn

register_exclusion(__file__)

T_co = TypeVar("T_co", covariant=True, bound=VarT, default=Any)

type UserValue[T: VarT = VarT] = VarRead[T] | T

Bool = UserValue[bool]
Int = UserValue[int]
# TODO: this causes pyright to report
# Type of parameter "x" is unknown (reportUnknownParameterType)
# on function "greater_than"
# when using --threads, sometimes
# type Float = UserValue[float]
type Float = VarRead[float] | float

type ValLabelLike = UserValue[Label] | str


class VarRead[T: VarT](abc.ABC):
    """
    covariant readonly base
    """

    @abc.abstractmethod
    def _read(self) -> Var[T]: ...

    def __check_co(self) -> VarRead[VarT]:  # pyright: ignore[reportUnusedFunction]
        return self

    __add__ = late_fn(lambda: add)
    __radd__ = late_fn(lambda: add)

    __lt__ = late_fn(lambda: less_than)
    __le__ = late_fn(lambda: less_than_or_eq)

    __gt__ = late_fn(lambda: greater_than)
    __ge__ = late_fn(lambda: greater_than_or_eq)


def _get[T: VarT](v: UserValue[T]) -> Value[T]:
    if isinstance(v, VarT):
        return v
    return v._read()


def _get_label(l: ValLabelLike) -> Value[Label]:
    if isinstance(l, str):
        return ensure_label(l)
    return _get(l)


def _get_type[T: VarT](v: UserValue[T]) -> type[T]:
    if isinstance(v, VarT):
        return type(v)
    if isinstance(v, Variable):
        return v._inner.type
    assert False


class Variable[T: VarT](VarRead[T]):
    """
    a MVar exposed to the user.

    This class is needed because Var/MVar have equality and compareison by id
    (needed in compiler internals); Variable equality will instead stage a equality operation.
    """

    _inner: MVar[T]

    def __init__(self, x: type[T] | UserValue[T]) -> None:
        if isinstance(x, type):
            self._inner = mk_mvar(x)
            return

        x_val = _get(x)
        self._inner = mk_mvar(get_type(x_val))
        self._inner.write(x_val)

    @staticmethod
    def _from_val[T1: VarT](v: Value[T1]) -> Variable[T1]:
        ans = Variable(get_type(v))
        ans._inner.write(v)
        return ans

    def _read(self) -> Var[T]:
        return self._inner.read()

    @property
    def value(self) -> Variable[T]:
        return Variable(self)

    @value.setter
    def value(self, v: UserValue[T]):
        self._inner.write(_get(v))


class Function[Ts: tuple[Any, ...]]:
    _instrs: Final[Ts]

    def __init__(self, *overloads: Unpack[Ts]) -> None:
        for x in overloads:
            assert isinstance(x, InstrBase)
        self._instrs = cast_unchecked(overloads)

    def __repr__(self) -> str:
        return "Function(" + ", ".join(type(x).__name__ for x in self._instrs) + ")"

    def __get__[V](self, obj: V, objtype: Any) -> _BoundFunction[V, Ts]:
        """
        typing helper to get type checker accept Function as a method
        for methods to actually work one must use for ex late_fn
        """
        assert False

    @overload
    def __call__[*I, O](
        self: Function[tuple[InstrTypedWithArgs_api[tuple[*I], O], *tuple[Any, ...]]], *args: *I
    ) -> O: ...
    @overload
    def __call__[*I, O](
        self: Function[tuple[Any, InstrTypedWithArgs_api[tuple[*I], O], *tuple[Any, ...]]],
        *args: *I,
    ) -> O: ...

    def __call__(self, *args: Any) -> Any:
        arg_types = tuple(_get_type(x) for x in args)

        for instr in self._instrs:
            assert isinstance(instr, InstrBase)
            if not can_cast_implicit_many(arg_types, instr.in_types):
                continue

            argvals = tuple(_get(x) for x in args)
            with must_use(), track_caller():
                ans = cast_unchecked(instr.call(*argvals))  # pyright: ignore
            if ans is None:
                return None
            assert isinst(ans, Var)

            return Variable._from_val(ans)

        raise TypeError()


class _BoundFunction[V, Ts: tuple[Any, ...]](Protocol):
    @overload
    def __call__[V1, *I, O](
        self: _BoundFunction[V1, tuple[InstrTypedWithArgs_api[tuple[V1, *I], O], *tuple[Any, ...]]],
        *args: *I,
    ) -> O: ...
    @overload
    def __call__[V1, *I, O](
        self: _BoundFunction[
            V1, tuple[Any, InstrTypedWithArgs_api[tuple[V1, *I], O], *tuple[Any, ...]]
        ],
        *args: *I,
    ) -> O: ...


################################################################################


@dataclass
class State[T = Any]:
    _scope: Scope
    _tree: pytree.PyTreeSpec | None = None
    _vars: list[MVar] | None = None

    def __init__(self, init: T | empty_t = empty):
        self._scope = _CUR_SCOPE.value
        if not isinstance(init, empty_t):
            self.write(init)

    def read(self) -> T:
        assert self._tree is not None
        assert self._vars is not None

        with track_caller():
            vars_ = [x.read() for x in self._vars]
        vars = [Variable._from_val(x) for x in vars_]
        return cast_unchecked(pytree.unflatten(self._tree, vars))

    def write(self, v: T):
        vars, tree = pytree.flatten(cast_unchecked(v))
        vars = [_get(x) for x in vars]

        if self._tree is None:
            assert self._vars is None
            with _CUR_SCOPE.bind(self._scope):
                self._tree = tree
                self._vars = [mk_mvar(get_type(x)) for x in vars]
        else:
            assert self._vars is not None
            assert tree == self._tree
            assert len(vars) == len(self._vars)
        with track_caller():
            for mv, arg in zip(self._vars, vars):
                mv.write(arg)


################################################################################


@dataclass
class TracedSubr[F = Any]:
    subr: RawSubr
    arg_tree: pytree.PyTreeSpec
    ret_tree: pytree.PyTreeSpec

    @property
    def call[**P, R](self: TracedSubr[Callable[P, R]]) -> Callable[P, R]:
        def inner(*args: P.args, **kwargs: P.kwargs) -> R:
            vals = self.arg_tree.flatten_up_to(cast_unchecked((args, kwargs)))
            vals_ = tuple(_get(cast_unchecked(x)) for x in vals)
            out_vars = self.subr.call(*vals_)

            out_vars_ = [Variable._from_val(x) for x in out_vars]
            return cast_unchecked(pytree.unflatten(self.ret_tree, out_vars_))

        return inner


_RETURN_HOOK: Cell[Callable[[Any], None]] = Cell()


def return_(val: Any = None) -> None:
    return _RETURN_HOOK.value(val)


def trace_to_subr[**P, R](
    fn: Callable[P, R | None], *args: P.args, **kwargs: P.kwargs
) -> TracedSubr[Callable[P, R]]:
    arg_vars, arg_tree = pytree.flatten(cast_unchecked((args, kwargs)))
    arg_types = tuple(_get_type(x) for x in arg_vars)

    out_tree: Cell[pytree.PyTreeSpec] = Cell()

    @functools.wraps(fn)
    def inner(*args: Var) -> tuple[Var, ...]:
        ar, kw = arg_tree.unflatten(Variable._from_val(x) for x in args)

        exit = mk_internal_label("trace_to_subr_ret")

        ret_state = State()

        def ret_hook(val: Any) -> None:
            ret_state.write(val)
            jump(exit)

        with _RETURN_HOOK.bind(ret_hook):
            ans = cast_unchecked(fn)(*cast_unchecked(ar), **cast_unchecked(kw))
            if ans is not None:
                with add_debug_info(
                    DebugInfo(describe=f"return val from end of {describe_fn(fn)}")
                ):
                    return_(ans)
            elif ret_state._tree is None:
                return_(None)

            _f.unreachable_checked()

        label(exit)

        assert ret_state._tree is not None
        assert ret_state._vars is not None
        out_tree.value = ret_state._tree
        return tuple(x.read() for x in ret_state._vars)

    ans = trace_to_raw_subr(arg_types, inner)

    return TracedSubr(ans, arg_tree, out_tree.value)


class Subr[F]:
    fn: Final[F]

    def __init__(self, fn: F) -> None:
        self.fn = fn
        self._subr: TracedSubr[F] | None = None

    @property
    def __call__[**P, R](self: Subr[Callable[P, R]]) -> Callable[P, R]:
        def inner(*args: P.args, **kwargs: P.kwargs) -> R:
            if self._subr is None:
                self._subr = trace_to_subr(self.fn, *args, **kwargs)
                # FIXME?
                # emit_frag(self._subr.subr.frag)
            return self._subr.call(*args, **kwargs)

        return inner


def subr():
    def inner[F](fn: F) -> Subr[F]:
        return Subr(fn)

    return inner


################################################################################

add = Function(AddI(), AddF())

unreachable_checked = Function(UnreachableChecked())


less_than = Function(PredLT())
less_than_or_eq = Function(PredLE())


def greater_than(x: Float, y: Float) -> Variable[bool]:
    return less_than(y, x)


def greater_than_or_eq(x: Float, y: Float) -> Variable[bool]:
    return less_than_or_eq(y, x)


def black_box[T: VarT](v: UserValue[T]) -> Variable[T]:
    return Function(BlackBox(_get_type(v)))(v)


################################################################################


def jump(label: ValLabelLike) -> None:
    return Jump().call(_get_label(label))


def branch(cond: Bool, on_true: ValLabelLike, on_false: ValLabelLike) -> None:
    return _f.branch(_get(cond), _get_label(on_true), _get_label(on_false))


def if_(cond: Bool) -> AbstractContextManager[None]:
    return trace_if(_get(cond))


################################################################################
