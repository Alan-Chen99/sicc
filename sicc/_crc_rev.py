# pyright: basic, reportOperatorIssue=false, reportAttributeAccessIssue=false, reportOptionalMemberAccess=false, reportWildcardImportFromLibrary=false

import zlib

from z3 import *

POLY = 0xEDB88320  # normal CRC-32 (IEEE 802.3)


# ----------------------------------------------------------------------
# Bit-precise CRC-32 expressed with Z3 bit-vectors
# ----------------------------------------------------------------------
def crc32_update(crc, byte):
    """one byte -> new 32-bit crc (all BitVec expressions)"""
    crc = crc ^ ZeroExt(24, byte)
    for _ in range(8):
        crc = If(crc & 1 == 1, LShR(crc, 1) ^ POLY, LShR(crc, 1))
    return crc


def crc32_expr(sym_bytes, start=0):
    """
    Build a Z3 expression for the CRC-32 of ‘sym_bytes’ starting from ‘start’
    (same convention as zlib.crc32).
    """
    crc = BitVecVal(start ^ 0xFFFFFFFF, 32)
    for b in sym_bytes:
        crc = crc32_update(crc, b)
    return crc ^ 0xFFFFFFFF


# ----------------------------------------------------------------------


def ascii_var(name):
    """8-bit symbolic byte"""
    return BitVec(name, 8)


def char_constraint(b):
    """b is 0-9 or a-z"""
    return Or(
        And(b >= ord("0"), b <= ord("9")),
        And(b >= ord("a"), b <= ord("z")),
        And(b >= ord("A"), b <= ord("Z")),
    )


def crc_rev(prefix: str, wanted_crc: int, length: int = 6) -> str | None:
    """
    find a string with a particular prefix and a crc,
    by appending a suffix of len `length`
    """
    crc_after_prefix = zlib.crc32(prefix.encode()) & 0xFFFFFFFF

    s = Solver()
    # symbolic characters of the suffix
    sym = [ascii_var(f"c{i}") for i in range(length)]
    # constrain alphabet
    for b in sym:
        s.add(char_constraint(b))

    # full crc expression
    crc_expr = crc32_expr(sym, start=crc_after_prefix)
    s.add(crc_expr == wanted_crc)

    if s.check() == sat:
        m = s.model()
        suffix = "".join(chr(m[b].as_long()) for b in sym)
        ans = prefix + suffix
        assert zlib.crc32(ans.encode()) == wanted_crc
        return ans

    raise RuntimeError()


# ----------------------------------------------------------------------
if __name__ == "__main__":
    for i in range(100):
        prefix = f"hello{i}_"
        target = 0x5AE7A6F2
        result = crc_rev(prefix, target)

        print(result or "no match within given length")
