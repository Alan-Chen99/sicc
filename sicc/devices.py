"""
this file will be replaced by a generated one
"""

from ._stationeers import DeviceBase
from ._stationeers import FieldDesc as _FieldDesc
from ._stationeers import mk_field as _mk_field


class _Lock:
    Lock: _FieldDesc[bool] = _mk_field(bool)


class _Open:
    Open: _FieldDesc[bool] = _mk_field(bool)


class Autolathe(DeviceBase, _Lock, _Open):
    pass


class AutomatedOven(DeviceBase, _Lock):
    pass
