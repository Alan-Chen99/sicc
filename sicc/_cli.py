from __future__ import annotations

import sys
from dataclasses import dataclass
from dataclasses import field
from typing import Annotated
from typing import Callable
from typing import Never

import cappa
from cappa.arg import Arg
from cappa.destructure import Destructured
from rich import print as print  # autoflake: skip
from rich.rule import Rule
from rich.text import Text

from ._api import loop
from ._diagnostic import describe_fn
from ._diagnostic import register_exclusion
from ._diagnostic import show_pending_diagnostics
from ._tracing import TracedProgram
from ._tracing import trace_program
from .config import Config
from .config import console_setup
from .config import with_rich_spinner
from .config import with_status

register_exclusion(__file__)


class SuppressExit(Exception):
    def __init__(self, code: int):
        self.code = code


@dataclass
class Program:
    fn: Callable[[], None]
    loop: bool

    def trace(self) -> TracedProgram:
        try:
            with trace_program() as res:
                if self.loop:
                    with loop():
                        self.fn()
                else:
                    self.fn()

            prog = res.value
            prog.check()
        finally:
            show_pending_diagnostics()
        return prog

    def cli(self) -> Never:
        cli = cappa.parse(Cli)
        cli.config.set_vars()
        console_setup()

        try:
            cli.call(self)
        except SuppressExit as e:
            exit(e.code)

        exit(0)


def program(loop: bool = False):
    def inner(fn: Callable[[], None]) -> Program:
        return Program(fn, loop=loop)

    return inner


group = cappa.Group(name="Global", section=1)


@cappa.command(name=sys.argv[0])
@dataclass
class Cli:
    config: Annotated[Config, Arg(destructured=Destructured(), hidden=True)] = field(
        default_factory=Config
    )
    cmd: cappa.Subcommands[Asm | Ir | Optimize | None] = None

    def call(self, prog: Program) -> None:
        if self.cmd is None:
            self.cmd = Asm()
        return self.cmd.call(prog, self)


@dataclass
class Ir:
    """
    print internal representation after a subset of optimizations
    """

    def call(self, prog: Program, cli: Cli) -> None:
        with with_rich_spinner(), with_status(f"{describe_fn(prog.fn)}"):
            ans = prog.trace()

        print(Rule(title=Text("Success; Internal Representation:", "ic10.title")))
        print(ans)


@dataclass
class Optimize:
    """
    print optimized version. instructions here have one-to-one correspondence to output asm
    """

    def call(self, prog: Program, cli: Cli) -> None:
        with with_rich_spinner(), with_status(f"{describe_fn(prog.fn)}"):
            f = prog.trace()
            f.optimize()

        print(Rule(title=Text("Success; Optimized:", "ic10.title")))
        print(f)


@dataclass
class Asm:
    """(default) run the compiler and get assembly"""

    def call(self, prog: Program, cli: Cli) -> None:
        with with_rich_spinner(), with_status(f"{describe_fn(prog.fn)}"):
            with with_status("Trace"):
                f = prog.trace()
            with with_status("Optimize"):
                f.optimize()
            with with_status("Regalloc"):
                f.regalloc()

        show_pending_diagnostics()

        print(Rule(title=Text("Success; Output (readable):", "ic10.title")))
        print(f)
        print(Rule(title=Text("Raw Equivalent:", "ic10.title")))
        print(f.gen_asm().text)
        print(Rule())
