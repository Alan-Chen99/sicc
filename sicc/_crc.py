from ._api import Int
from ._api import Str
from ._api import Variable
from ._api import if_
from ._api import range_
from ._api import subr


@subr
def crc_append(s: Str, b: Int) -> Str:
    crc_ = Variable(s).transmute(int)

    # clear the left 32 bits
    # (they are all 1 if the hash is given as negative)
    crc_ &= 2**32 - 1

    crc_ ^= 0xFFFFFFFF
    crc_ ^= b

    crc = Variable(crc_)

    for _idx in range_(8):
        right_bit = crc & 1
        crc.value >>= 1
        with if_(right_bit != 0):
            crc.value ^= 0xEDB88320

    ans = Variable(crc ^ 0xFFFFFFFF)

    # normalize back as negative
    with if_(ans >= 2**31):
        ans.value -= 2**32

    return ans.transmute(str)
