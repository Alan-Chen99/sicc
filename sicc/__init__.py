# autoflake: skip_file

# modules
from . import devices as devices
from . import functions as functions

# high level
from ._cli import Program as Program
from ._cli import program as program
from ._diagnostic import show_pending_diagnostics as show_pending_diagnostics

# types
from ._api import UserValue as UserValue
from ._api import Variable as Variable
from ._core import Label as Label
from ._core import nan as nan
from ._stationeers import BatchMode as BatchMode
from ._stationeers import Color as Color
from ._stationeers import LogicType as LogicType

# UserValue aliases
from ._api import Bool as Bool
from ._api import Float as Float
from ._api import Int as Int
from ._api import Str as Str
from ._api import ValLabelLike as ValLabelLike
from ._stationeers import ValBatchMode as ValBatchMode
from ._stationeers import ValLogicTypeLike as ValLogicTypeLike

# control flow primitives
from ._api import asm as asm
from ._api import asm_block as asm_block
from ._api import asm_fn as asm_fn
from ._api import branch as branch
from ._api import cond as cond
from ._api import jump as jump
from ._api import mk_label as mk_label
from ._tracing import exit_program as exit_program
from ._tracing import label as label

# control flow high level
from ._api import cjump as cjump
from ._api import if_ as if_
from ._api import loop as loop
from ._api import while_ as while_
from ._tracing import break_ as break_
from ._tracing import continue_ as continue_
from ._tracing import else_ as else_

# subroutines
from ._api import return_ as return_
from ._api import subr as subr

# dataclasses
from ._tree_utils import dataclass as dataclass
from ._tree_utils import dataclasses as dataclasses
from ._tree_utils import pytree as pytree

field = dataclasses.field  # pyright: ignore[reportUnknownMemberType, reportUnknownVariableType]

# devices
from ._stationeers import Device as Device
from ._stationeers import DeviceBase as DeviceBase
from ._stationeers import DeviceTyped as DeviceTyped
from ._stationeers import Pin as Pin
from ._stationeers import pin as pin

db = pin("db")
d0 = pin(0)
d1 = pin(1)
d2 = pin(2)
d3 = pin(3)
d4 = pin(4)
d5 = pin(5)


# functions
from ._api import comment as comment
from ._api import select as select
from ._api import undef as undef
from ._stationeers import yield_ as yield_
