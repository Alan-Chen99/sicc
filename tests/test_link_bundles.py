from sicc import *
from sicc.devices import *
from test_utils import wrap_test

d = Autolathe()


@subr()
def func(x: Int) -> None:
    d.Activate = x > 3


@wrap_test
@program()
def test_cond_call():
    """currently function that returns dont work. will hopefully be fixed latter"""
    with if_(d.On):
        func(d.ImportCount)

    with if_(d.Power == 2):
        with else_():
            func(d.ImportCount)


@dataclass
class Data:
    x: Float
    y: Float


@subr()
def child(x: Float):
    with if_(x > 0):
        return_(Data(7, x))
    return Data(x, 9)


@subr()
def parent(x: Float) -> Float:
    with if_(x > 0):
        return_(child(x).x)

    return x + 5


@wrap_test
@program()
def test_nested_call():
    comment("child", child(1).x)
    comment("res1", parent(1))
    comment("res2", parent(2))


dev_name = Variable(str)
ret_to = Variable(Label)


@subr()
def map_devices(func_ptr: ValLabelLike):
    for d in range(5):
        ret_l = mk_label(f"process Autolathe{d} ret")

        dev_name.value = f"Autolathe{d}"
        ret_to.value = ret_l
        jump(func_ptr)
        label(ret_l)


@wrap_test
@program()
def test_function_ptr():

    with if_(False):

        fn1 = label("fn1")
        Autolathe(dev_name).On = True
        jump(ret_to)

        fn2 = label("fn2")
        with if_(Autolathe(dev_name).On):
            Autolathe(dev_name).Activate = True
        jump(ret_to)

    map_devices(fn1)
    map_devices(fn2)
