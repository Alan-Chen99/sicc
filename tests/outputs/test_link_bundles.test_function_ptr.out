move r1 10
move r2 3
j 12
move r2 23
move r1 6
j 12
lbn r0 336213101 r3 On 0
blt r0 1 9
sbn 336213101 r3 Activate 1
j ra
sbn 336213101 r3 On 1
j ra
move r3 982299349
jal r1
move r3 1300988483
jal r1
move r3 -729623559
jal r1
move r3 -1551506577
jal r1
move r3 1038186188
jal r1
j r2
# noop (jump here to exit)
# 24 lines total (not including this line)
